### PR TITLE
Add function time logger

### LINKED ISSUE #2542: Log time of different steps in analysis step
We should log roughly how long the analysis step takes, and the ministep as this would be helpful when refactoring. Must be done after https://github.com/equinor/ert/pull/2500 is merged.

### PR DESCRIPTION
**Pull request by sregales-TNO**
**Issue**
Resolves #2542 

**Approach**
Create a function that takes another function as an argument and log that function execution time

### CODE REVIEW

**Comment by dotfloat** [comment]
I think you should use RAII rather than making a wrapper function:

```c++
class Benchmark {
using clock = std::chrono::high_resolution_clock;

clock::time_point m_start_time = clock::new();
std::string m_message;
std::shared_ptr<ILogger> m_logger;

public:
Benchmark(std::shared_ptr<ILogger> logger, const std::string &message):
    m_logger(logger),
    m_message(message),
{}

~Benchmark() {
    m_logger->info("Time taken by '{}': {}", m_message, clock::now() - m_start_time);
}
};
```

Usage:
```c++
namespace {
auto logger = ert::get_logger("mylogger");
}

void function_you_want_to_bench() {
    Benchmark benchmark{logger, __PRETTY_FUNCTION__};
    // lots of stuff
    // <- `benchmark` goes out of scope and logs time to Python
}
```

Add `#include <fmt/chrono.h>` to `ert/python.hpp`.

Refer to fmtlib docs for how to use the `fmt/chrono.h`: https://fmt.dev/latest/api.html#date-and-time-formatting

You can also turn this into a macro:
```c++
#define BENCHMARK() \
    Benchmark _benchmark{logger, __PRETTY_FUNCTION__}
```

Then:
```c++
void function_you_want_to_bench() {
    BENCHMARK();  // this assumes there's a variable named `logger` in scope
    // lots of stuff
    // <- `benchmark` goes out of scope and logs time to Python
}
```

**Comment by sregales-TNO** [comment]
I see we have the following include```<ert/util/*.hpp>```, but I'm unable to find them in the include directory. I would like to make this live there

**Comment by dotfloat** [comment]
`ert/util` is located in libecl. You can make a directory in libres' `include` if you want.

**Comment by oyvindeide** [comment]
> `ert/util` is located in libecl. You can make a directory in libres' `include` if you want.

There is already a `res_utils`, might want to put it there?But now I am bikeshedding.

File: ert3/stats/_stats.py [hunk 1]
```diff
@@ -47,8 +47,8 @@ def _to_record(self, x: np.ndarray) -> ert.data.Record:  # type: ignore
             return ert.data.NumericalRecord(data=x.tolist())
         else:
             return ert.data.NumericalRecord(
-                data={
-                    idx: float(val) for idx, val in zip(self.index, x)  # type: ignore
+                data={  # type: ignore
```
Comment (xjules) [inline ert3/stats/_stats.py:50]: Why's the difference here? Another black version?

File: libres/lib/include/ert/utils/metric.hpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        : m_logger(logger), m_message(message) {}
+
+    ~Benchmark() {
+        m_logger->info(
+            "{} execution time: {:.4f} s", m_message,
+            std::chrono::duration<float>(clock::now() - m_start).count());
+    }
+};
+} // namespace utils
+} // namespace ert
```
Comment (xjules) [inline libres/lib/include/ert/utils/metric.hpp:25]: Missing new line?

File: libres/lib/include/ert/utils/metric.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    std::shared_ptr<ILogger> m_logger;
+
+public:
+    Benchmark(std::shared_ptr<ILogger> logger, const std::string &message)
+        : m_logger(logger), m_message(message) {}
+
+    ~Benchmark() {
+        m_logger->info(
+            "{} execution time: {:.4f} s", m_message,
+            std::chrono::duration<float>(clock::now() - m_start).count());
```
Comment (xjules) [inline libres/lib/include/ert/utils/metric.hpp:21]: When I was measuring some GPU computation this's what I had used (not sure what's the difference though):
```
auto rtime = std::chrono::duration_cast<std::chrono::microseconds>(end_time - begin_time).count()
```

File: libres/lib/include/ert/python.hpp [hunk 4]
```diff
@@ -61,6 +61,7 @@ class ILogger {
  * is called, and before Python shuts down, but is valid for all other times.
  */
 std::shared_ptr<ILogger> get_logger(const std::string &name);
+
```
Comment (xjules) [inline libres/lib/include/ert/python.hpp:64]: Remove the new line here I guess.

File: ert3/stats/_stats.py (hunk 1 above)
Comment (sregales-TNO) [reply ert3/stats/_stats.py:50]: mypy was complaining on github workflow typing. speculating here but I think it had to do with yesterday's bumped version, which has already been fixed. will try it again with mypy old version.

File: libres/lib/include/ert/utils/metric.hpp (hunk 3 above)
Comment (sregales-TNO) [reply libres/lib/include/ert/utils/metric.hpp:21]: Duration is a class and duration_cast is a function that cast the incoming duration to another `(end_time - begin_time)` is implicitly a duration with a rep `unsigned int type` with a period `ratio<1,1>` aka [std::chrono::seconds](https://www.cplusplus.com/reference/chrono/seconds/)
so another way that yours could be written is 
```cpp
auto rtime = std::chrono::duration_cast<std::chrono::duration<int, std::mirco>>(end_time - begin_time).count()
//so since I wanted float and seconds, if I wanted to use duration_cast it would of been redundant example
auto rtime = std::chrono::duration_cast<std::chrono::duration<float, std::ratio<1,1>>>(end_time - begin_time).count()
```
so instead, I opted to use the class directly

File: ert3/stats/_stats.py (hunk 1 above)
Comment (sregales-TNO) [reply ert3/stats/_stats.py:50]: weird, the problem still persists, so how did it pass when it first got written?

Comment (sregales-TNO) [reply ert3/stats/_stats.py:50]:
```bash
❯ mypy --version
mypy 0.910
❯ mypy -p ert -p ert3
ert/data/record/_record.py:22: error: Module "beartype.roar" has no attribute "BeartypeDecorHintPepDeprecationWarning"; maybe "BeartypeDecorHintPepDeprecatedWarning", "BeartypeDecorHintPepWarning", or "BeartypeDecorHintPepException"?
ert3/stats/_stats.py:50: error: Argument "data" to "NumericalRecord" has incompatible type "Dict[object, float]"; expected "Union[List[Union[int, float]], Dict[str, Union[int, float]], Dict[int, Union[int, float]]]"
Found 2 errors in 2 files (checked 42 source files)
```

File: libres/lib/include/ert/python.hpp (hunk 4 above)
Comment (sregales-TNO) [reply libres/lib/include/ert/python.hpp:64]: I think the terminating `newline` here and the thereof in `metric.hpp`, shows some inconsistency when it comes to clang-format

File: libres/lib/include/ert/utils/metric.hpp (hunk 3 above)
Comment (xjules) [reply libres/lib/include/ert/utils/metric.hpp:21]: I see, the default is then the second precision which is fine I guess then :)

File: ert3/stats/_stats.py (hunk 1 above)
Comment (xjules) [reply ert3/stats/_stats.py:50]: Right, the thing self.index can be of two types in `NumericalRecord`; ie. `RecordIndex = Union[Tuple[int, ...], Tuple[str, ...]]` 
The best approach would be to check `isinstance(self.index[0], str)` and `isinstance(self.index[0], int)` I reckon ...

Comment (xjules) [reply ert3/stats/_stats.py:50]: I also understand that this is not part of your PR. Not sure what's the best approach.

Comment (sregales-TNO) [reply ert3/stats/_stats.py]: this issue is being handled here: #2642

File: libres/lib/include/ert/python.hpp (hunk 4 above)
Comment (oyvindeide) [reply libres/lib/include/ert/python.hpp:64]: I could remove the newline on my computer and `clang-format` did not complain?

**Comment by BjarneHerland** [comment]
I see that using a macro was suggested earlier in the conversation but I fail to see the value in this case. IMO it is clearer and more tidy to just use the class.

A macro makes it simple to enable/disable profiling on a "global scale" (by just defining the macro to empty and recompile). We do, however, not deal with time-critical code in this case and our goal is not profiling/optimizing pieces of the code, but real-life telemetry to gain insight. We will most likely enable/disable such telemetry on a case-by-case basis and not globally anyway. **_Note that I'm not being dogmatic about this - just my opinion - so feel free to ignore._**

Apart from this I like the solution and would approve it.

**Comment by pinkwah** [comment]
> I see that using a macro was suggested earlier in the conversation but I fail to see the value in this case. IMO it is clearer and more tidy to just use the class.
> 
> A macro makes it simple to enable/disable profiling on a "global scale" (by just defining the macro to empty and recompile). We do, however, not deal with time-critical code in this case and our goal is not profiling/optimizing pieces of the code, but real-life telemetry to gain insight. We will most likely enable/disable such telemetry on a case-by-case basis and not globally anyway. **_Note that I'm not being dogmatic about this - just my opinion - so feel free to ignore._**
> 
> Apart from this I like the solution and would approve it.

My suggestion to have a macro is because the variable itself is of never going to be used. So we'd either have `Benchmark _benchmark(logger);` or worse: `Benchmark(logger);`. The latter constructs and deconstructs the instance immediately, so it's a benchmark of `Benchmark` itself, but is valid and natural code that the code reviewer will have to catch.

Therefore, I think the use-case is clearer with `BENCHMARK();`

Note also the `__PRETTY_FUNCTION__` macro is used. The alternative would be that the programmer duplicates the name of the function/forgets to change it.

Tl;dr: Imo, fewer ways it can go wrong with `BENCHMARK();`.

**Comment by BjarneHerland** [comment]
As mentioned, I'm not going to argue on the point of macro. 😄  

Personally, I would've trusted the programmer to understand the concept of RAII (more presicely SBRM in this case) and exposed to her the fact that this utilized. But again - it's more a matter of taste than right or .wrong.

**Comment by sregales-TNO** [comment]
based on the conversation I removed the use of macros and while fixing code conflicts I adjusted my code to be more nsync with the memory module.

**Comment by BjarneHerland** [comment]
Some simple test would be nice, perhaps along the lines of the [memory module test](https://github.com/equinor/ert/blob/main/libres/tests/res_util/test_memory.cpp)

(You just need the MockLogger-part, ignoring the "mocking" of `get_file()` )

**Comment by sregales-TNO** [comment]
test this please

File: libres/tests/logger_mock.hpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#include <fmt/format.h>
```
Comment (pinkwah) [inline libres/tests/logger_mock.hpp:1]: This is a header file. Add a `#pragma once` at the top.

**Review by pinkwah — APPROVED** [review]
